### PR TITLE
Implement queryRenderedFeatures for heatmap layers

### LINKED ISSUE #9112: map.on('mouseenter', ...) doesn't work for heatmap layer
**mapbox-gl-js version**: 1.6.0

**browser**: firefox developer edition

### Steps to Trigger Behavior

 1. hover over points to see that there is a popup on mouseenter
 2. change map.on('mouseenter', 'earthquakes-point',...) to map.on('mouseenter', 'earthquakes-heatmap',...)
 3. hover over heatmap layer to see that the popup is no longer working

### LINKED ISSUE #10957: Add queryRenderedFeatures support to heatmap layer
Currently the heatmap layer doesn't support `queryRenderedFeatures` — the latter simply doesn't return anything for such layers: https://github.com/mapbox/mapbox-gl-js/blob/12968ab002c00fffbbd255df5604493b6fa6dfb4/src/style/style_layer/heatmap_style_layer.js#L65-L67

I always assumed we would need to implement difficult raster querying logic for layers like this to expose "intensity under the query point", but it looks like users actually don't need this in most cases. What's usually needed is simply "check whether there's a heatmap under the cursor, yes/no" — and this is much easier to implement. For example use cases, see #9112, and currently https://github.com/mapbox/studio/issues/10803 for selecting heatmap layers in Studio.

We can implement this by querying the underlying data for nearby features that can influence the query point — specifically, reusing the [circle layer's `queryIntersectsFeature` logic](https://github.com/mapbox/mapbox-gl-js/blob/main/src/style/style_layer/circle_style_layer.js#L47) and substituting `circle-*` properties for equivalent `heatmap` ones. So, `heatmap-radius` instead `circle-radius` + `circle-stroke-width`, assume `circle-pitch-alignment` and `circle-pitch-scale` as equal to `map`, etc.

### PR DESCRIPTION
**Pull request by mourner**
Closes #10957. Closes #9112. Implements `queryRenderedFeatures` support for heatmap layers by reusing the logic from the circle layer. Instead of always returning `false`, it now returns an event containing all features that influence the heatmap rendering in the queried geometry (which are essentially features within the `heatmap-radius` of it). Makes the following possible:

```js
map.on('click', 'heatmap', (e) => {
  console.log(`Clicked the heatmap! Features influencing this point: ${e.features.length}`);
});
```

cc @samanpwbb @MaretIdris @ivovandongen 

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [x] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [x] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Added queryRenderedFeatures support to heatmap layers.</changelog>`

### CODE REVIEW

**Review by rreusser — APPROVED** [review]
I gave this a test run and looked it over pretty carefully, and… it looks really good. 👏  Could possibly add a test, but not sure if that's worthwhile.

As for documentation a list of places where we could note this:
- [heatmap layer docs](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#heatmap): perhaps an odd place, though people probably end up here if they want to create a heatmap.
- [Create a heatmap layer example](https://docs.mapbox.com/mapbox-gl-js/example/heatmap-layer/): Could add even just an output that shows how many points were clicked so that people know they *can* query this and sort through the data, if desired.
- [click event docs](https://docs.mapbox.com/mapbox-gl-js/api/map/#map.event:click): it doesn't currently mention these details, but it could be helpful to give a verbose list of what's outputted in the event data or which layers are supported.

Though it's quite possible I've missed a more appropriate place for this documentation.

**Comment by mourner** [comment]
> Could possibly add a test, but not sure if that's worthwhile.

Yep, added query tests at the same time as you commented. 👍 

> Though it's quite possible I've missed a more appropriate place for this documentation.

Perhaps worth adding a sentence in the description of [heatmap-radius](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#paint-heatmap-heatmap-radius), noting that `queryRenderedFeatures` will return features within this radius from the query point?

**Comment by rreusser** [comment]
> Perhaps worth adding a sentence in the description of heatmap-radius

Yes, that seems like a very reasonable place. I think someone interested in this subtlety will likely end up reading the style documentation pretty carefully. 👍

**Comment by arindam1993** [comment]
Looks good to me. 

I think there also would be a lot of value in adding the heatmap weight/color to the response, atleast for single point queries. 
If we want to get fancy we could do it for box-queries too with some FBO rendering and `gl.readPixels()`, but probably not worth the complexity.

**Review by arindam1993 — APPROVED** [review]

File: src/style/style_layer/heatmap_style_layer.js [hunk 1]
```diff
@@ -58,12 +67,23 @@ class HeatmapStyleLayer extends StyleLayer {
...
+                           geometry: Array<Array<Point>>,
+                           zoom: number,
+                           transform: Transform,
+                           pixelPosMatrix: Float32Array,
+                           elevationHelper: ?DEMSampler): boolean {
+
+        const size = this.paint.get('heatmap-radius').evaluate(feature, featureState);
+        return queryIntersectsCircle(
+            queryGeometry, geometry, transform, pixelPosMatrix, elevationHelper,
+            true, true, new Point(0, 0), size);
```
Comment (arindam1993) [inline src/style/style_layer/heatmap_style_layer.js:86]: TIL we don't have `heatmap-translate` I thought it was feature available across all layers.

**Review by arindam1993 — APPROVED** [review]

**Comment by mourner** [comment]
> I think there also would be a lot of value in adding the heatmap weight/color to the response, atleast for single point queries.

Possibly, and previously I assumed this would be required for implementing heatmap querying, but all use cases I encountered so far didn't need this, as described in #10957. Then there's this conceptual problem of what the density value would represent — in the context of interactive GL JS maps, it's used purely for visual purposes, so heatmap density or color at a particular point isn't very meaningful. So I suggest leaving this for future if use cases like this do arise.

**Pull request merged**